index: Ensure restriction is supported in find_idx

dekimir opened the pull request:
Previously, statement_restrictions::find_idx() would happily return an
index for a non-EQ restriction (because it checked only the column
name, not the operator).  This is incorrect: when the selected index
is for a non-EQ restriction, it is impossible to query that index
table.

Fixes #7659.

Tests: unit (dev)

psarna reviewed (approved):
Ouch, a sneaky bug. Looks good.

avikivity commented on cql3/restrictions/statement_restrictions.cc:
```diff
@@ -329,20 +329,38 @@ int statement_restrictions::score(const secondary_index::index& index) const {
+        return partition->restrictions();
+    } else if (auto clustering = dynamic_cast<const single_column_clustering_key_restrictions*>(rs)) {
+        return clustering->restrictions();
+    }
```
This code is saying, please convert me to expr, I'll be so much nicer.

avikivity replied:
(of course this quick fix should come first)

avikivity commented:
I have a question about `is_supported_by()`:

```c++
bool is_supported_by(const expression& expr, const secondary_index::index& idx) {
    using std::placeholders::_1;
    return std::visit(overloaded_functor{
            [&] (const conjunction& conj) {
                return boost::algorithm::all_of(conj.children, std::bind(is_supported_by, _1, idx));
            },
            [&] (const binary_operator& oper) {
                return std::visit(overloaded_functor{
                        [&] (const column_value& col) {
                            return idx.supports_expression(*col.col, oper.op);
                        },
                        [&] (const std::vector<column_value>& cvs) {
                            return boost::algorithm::any_of(cvs, [&] (const column_value& c) {
                                return idx.supports_expression(*c.col, oper.op);
                            });
                        },
                        [&] (const token&) { return false; },
                    }, oper.lhs);
            },
            [] (const auto& default_case) { return false; }
        }, expr);
}
```

We use all_of() for a conjunction, but any_of() of vector<column_value>. I think any_of is correct for both; if an index supports one restriction, the other can be filtered. That's better than filtering everything.

`WHERE a = :a AND b = :b` -> should match an index on a, but doesn't now.

`WHERE (a, b) = :a_b` -> matches an index on a, as it should

If I'm correct, then this change introduces a regression, and is_supported_by() should be fixed first.




(looks like it wasn't a question but a statement)

dekimir commented:
> `WHERE a = :a AND b = :b` -> should match an index on a, but doesn't now.

It does, as each atom goes into a separate map entry.

> WHERE (a, b) = :a_b -> matches an index on a, as it should

This is independent of the any/all_of behaviour of conjunction.

I tried switching conjunction to any_of in is_supported_by(), but then the following case bombs: `WHERE a>10 AND a=11`.  The rest of the code isn't ready to deal with this yet.

> If I'm correct, then this change introduces a regression, and is_supported_by() should be fixed first.

Why precisely do you think a regression is introduced?  I'm not seeing it.

nyh commented on test/boost/secondary_index_test.cc:
```diff
@@ -255,6 +255,9 @@ SEASTAR_TEST_CASE(test_many_columns) {
         e.execute_cql("CREATE INDEX ON tab (e)").get();
         e.execute_cql("CREATE INDEX ON tab (f)").get();
+        // #7659
+        require_rows(e, "SELECT * FROM tab WHERE d=0 AND f>0 ALLOW FILTERING", {});
```
Don't you need eventually() here because you didn't wait for the index creation to finish?
Did this test fail before this patch?

dekimir replied:
> Don't you need eventually() here because you didn't wait for the index creation to finish?

Apparently, _I_ don't, but that may be just a quirk of my dev environment.  I think I'll add it anyway.

> Did this test fail before this patch?

Yes, it failed in the same way as #7659.

nyh reviewed (commented):
Looks almost good to me, but isn't the test missing an eventually()?

dekimir commented:
> > WHERE (a, b) = :a_b -> matches an index on a, as it should
> 
> This is independent of the any/all_of behaviour of conjunction.

Please ignore this remark; I just realized why you gave that example...

The rest of my comment still stands.  I'm not sure I see any regressions?

avikivity commented:
> > `WHERE a = :a AND b = :b` -> should match an index on a, but doesn't now.
> 
> It does, as each atom goes into a separate map entry.

What do you mean? that these are two expressions? What's the map?

If indeed it turns out that these are two expressions, there's still a bug here, it's just latent.

> 
> > WHERE (a, b) = :a_b -> matches an index on a, as it should
> 
> This is independent of the any/all_of behaviour of conjunction.
> 
> I tried switching conjunction to any_of in is_supported_by(), but then the following case bombs: `WHERE a>10 AND a=11`. The rest of the code isn't ready to deal with this yet.

Why does it bomb? just curious.

> 
> > If I'm correct, then this change introduces a regression, and is_supported_by() should be fixed first.
> 
> Why precisely do you think a regression is introduced? I'm not seeing it.

If `WHERE a = :a AND b = :b` is a single expression, then it wouldn't match an index on a due to all_of.

If it's not a single expression, then it's not a regression, just a latent bug that is not expressed.

dekimir commented:
> > > `WHERE a = :a AND b = :b` -> should match an index on a, but doesn't now.
> > 
> > 
> > It does, as each atom goes into a separate map entry.
> 
> What do you mean? that these are two expressions? What's the map?

See `rmap` in the changed code.

> > I tried switching conjunction to any_of in is_supported_by(), but then the following case bombs: `WHERE a>10 AND a=11`. The rest of the code isn't ready to deal with this yet.
> 
> Why does it bomb? just curious.

IIRC, `used_index_restrictions->value_for()` doesn't find a single value and asserts out.  I'll recheck when convenient, right now I'm in the middle of something else.

> If `WHERE a = :a AND b = :b` is a single expression, then it wouldn't match an index on a due to all_of.
> 
> If it's not a single expression, then it's not a regression, just a latent bug that is not expressed.

It's not a single expression.  There may be a bug, but it cannot be exposed by any query until we take the front-end rewrite a step further.

PS1: Please don't merge yet, I discovered a bug in how multi-column restrictions are treated.

PS2: The 4.2 fix will likely diverge further from this patch.  Do you think I should submit a separate patch for 4.2?

dekimir commented:
OK, the [bug I discovered](#7680) is orthogonal.  I addressed feedback and added more tests.  [Changelog](https://github.com/dekimir/scylla/commits/fix-find_idx-ineq.v2):
- 3435e3be8 Handle multi-column restrictions in find_idx()
- 6735b0e8c Use eventually() in new tests
- 809d10870 Add testcase for multiple atoms on same column

avikivity commented:
> 
> PS2: The 4.2 fix will likely diverge further from this patch. Do you think I should submit a separate patch for 4.2?

Focus on fixing on master. My approach is to prepare a backportable fix if it is natural - if the code did not diverge too much. If it did, then priority is to keep master clean and moving in the correct direction, and if it means more backport work, then that's what I do.

avikivity commented:
> > > > `WHERE a = :a AND b = :b` -> should match an index on a, but doesn't now.
> > > 
> > > 
> > > It does, as each atom goes into a separate map entry.
> > 
> > 
> > What do you mean? that these are two expressions? What's the map?
> 
> See `rmap` in the changed code.
> 

I see. And it's already stored split.

> > > I tried switching conjunction to any_of in is_supported_by(), but then the following case bombs: `WHERE a>10 AND a=11`. The rest of the code isn't ready to deal with this yet.
> > 
> > 
> > Why does it bomb? just curious.
> 
> IIRC, `used_index_restrictions->value_for()` doesn't find a single value and asserts out. I'll recheck when convenient, right now I'm in the middle of something else.
> 
> > If `WHERE a = :a AND b = :b` is a single expression, then it wouldn't match an index on a due to all_of.
> > If it's not a single expression, then it's not a regression, just a latent bug that is not expressed.
> 
> It's not a single expression. There may be a bug, but it cannot be exposed by any query until we take the front-end rewrite a step further.

Ok, so at  least it wants a comment.

dekimir commented:
Pausing this until #7699 is merged.  That will "fix" is_supported_by and allow me to simplify this patch.

dekimir commented:
Redone after #7699, PTAL.  [Changelog](https://github.com/dekimir/scylla/commits/fix-find_idx-ineq.v3):
* 8a25356fe Don't copy restrictions_map in get_individual_restrictions_map
* d4c4ef94a Revert 3435e3b Handle multi-column restrictions in find_idx()
* rebased on master
